diff --git a/lua/codecompanion/interactions/inline/keymaps.lua b/lua/codecompanion/interactions/inline/keymaps.lua
index ccd1a343..c17244ce 100644
--- a/lua/codecompanion/interactions/inline/keymaps.lua
+++ b/lua/codecompanion/interactions/inline/keymaps.lua
@@ -19,7 +19,7 @@ end
 M.stop = {
   callback = function(inline)
     inline:stop()
-    clear_map(config.interactions.inline.keymaps, inline.diff.bufnr)
+    clear_map(config.interactions.inline.keymaps, inline.bufnr)
     log:trace("[Inline] Cancelling the request")
   end,
 }
