diff --git a/lua/codecompanion/strategies/chat/ui/init.lua b/lua/codecompanion/strategies/chat/ui/init.lua
index 3d8957f..1bc7dd7 100644
--- a/lua/codecompanion/strategies/chat/ui/init.lua
+++ b/lua/codecompanion/strategies/chat/ui/init.lua
@@ -71,7 +71,9 @@ function UI:open(opts)
     return
   end
   if config.display.chat.start_in_insert_mode then
-    vim.cmd("startinsert")
+    vim.schedule(function()
+      vim.cmd("startinsert")
+    end)
   end
 
   local window = config.display.chat.window
